File: arrays/QuickSort.ts
function qs(arr: number[], lo: number, hi:number): void{
//calls parttition, does the pivot and does the recurssive step
//In QS hi in inclusive which isnt normal in algo 
    if (lo >= hi){
        return
    }

    const pivotIdx = partition(arr, lo, hi);

    qs(arr, lo, pivotIdx -1);
    qs(arr, lo, pivotIdx + 1, hi); 
}

function partition(arr: number[], lo:number, hi:number): number {
//gets the pivot and moves the items to their correct side of the pivot


    //defines the new pivot for each subarray and makes the arrays gradually more sorted as it removes elements from its immediate scope
    const pivot = arr[hi];

    
    let idx = lo -1;
    
    //walks from low to hi but not the pivot, and everything is being compared to the pivot 
    for (let i = lo; i < hi; ++i){
        //if you find an element which is smaller than or equal to the pivot move it to the left of the pivot and increment the index. 
        //eventually everything smaller than the pivot will be to the left of it, and everything larger will be to the right of it 
        if (arr[i] <= pivot){
            //incremenets the index to move along the array 
            idx++;
            const tmp = arr[i];
            arr[i] = arr[idx];
            arr[idx] = tmp;
        } 
    }
    //puts the pivot into the position where = [vals < p > vals]
    idx++;
    arr[hi] = arr[idx];
    arr[idx] = pivot;

    return idx
}

export default function quick_sort(arr: number[]): void{
    qs(arr, 0, arr.length - 1);
    
}

File: arrays/QuickSort.py
def qs(arr, lo, hi):
    if  lo < hi:
     
        #gets pivot index in middle of arr
        pivotIdx = partition(arr, lo, hi)

        #this runs to branches to the right(higher) and left (lower) of the pivot creating 
        #more branches until eventually almost every value has been the pivot and sorted
        #or was one of the remaining values that got sorted left or right of the final pivot
        qs(arr, lo, pivotIdx - 1)
        qs(arr, pivotIdx +1, hi)

def partition(arr, lo, hi):
    
    #make the pivot the last element in the array
    pivot = arr[hi]

    #idex before the array to find the first time a number is less than the pivot and make sure its 
    #in the first element in the array at the begining
    idx = lo -1

    for i in range(lo, hi):
        if arr[i] <= pivot:
            #this moves the index initially to the first element in the aray and will swap the first 
            #element it finds that is smaller than the pivot
            idx = idx +1 
            #swaps the elements
            arr[i], arr[idx] = arr[idx], arr[i]
    
    idx = idx +1
    arr[hi], arr[idx] = arr[idx], arr[hi]
    

    return idx



arr = [9,4,5,3,2,7,9,3,2,5,66,5,34,4,43,2,5,1]
size = len(arr)
ss = qs(arr, 0, (size - 1))
print(arr)